packages/@aws-cdk/aws-elasticloadbalancingv2/lib/alb/application-listener-rule.ts

```diff
@@ -160,6 +160,227 @@ export interface RedirectResponse {
   readonly statusCode: string;
 }
 
+/**
+ * Properties for a raw Cfn listener rule condition
+ */
+export interface RawListenerRuleConditionProps {
+  readonly [key: string]: any;
+}
+
+/**
+ * Interface for a listener rule condition
+ */
+export interface IListenerRuleCondition {
```

Comment:
What you're doing here is good, but what we generally do is:

```ts
export [abstract] class ListenerRuleCondition {
  public static queryString(key, value): ListenerRuleCondition {
    return new ...;
  }

  public static hostHeader(values: string[]): ListenerRuleCondition {
    return new ...;
  }

  
  protected constructor(...) { ... }
}

// NOTE: not "export"ed.
class IfYouNeedThem extends ListenerRuleCondition {
}
```

So:

* (abstract) base class instead of interface. Whether it's abstract or not depends on how you want to model it. It's perfectly fine to make the class itself just a holder for a blob of JSON that it gets in the constructor (which is protected) and spits back out in the `render()` method for like 90% of the cases. If it turns out you need a specific subclass for some other case, you still can do that. Also still leaves the ability for customers to define their own subclasses.
* Static factory methods to create instances.
* (Generally) hide the specific implementation classes (if you have any).

This has the advantage that it's more discoverable. The declared type `conditions?: ListenerRuleCondition[]` will also be the type that you can just type into your IDE, hit `.`, and then see an autocompleted list of options. Otherwise you'll have to go digging around for all the classes you can `new` up, which is more work.

It's also fine to have two or more factory functions for the same type of condition, if their signatures are slightly different. Then you can make "simple" and convenient ones, and more complex ones, and don't have to expose the full complexity of the API to people who just need a simple condition.

Also wondering if we can just call this class `Condition` instead of `ListenerRuleCondition`, since people are going to have to type it a lot :/

Reply:
Maybe also stick all the conditions in their own file.

Reply:
That's better! I'll refactor as this!